R/new-blog-post.R

```diff
@@ -0,0 +1,82 @@
+#' Create a new blog post
+#'
+#' Creates (and potentially opens) the `index.qmd` file for a new blog post.
+#'
+#' @param title A character string for the title of the post. It is converted
+#' to title case via [tools::toTitleCase()].
+#' @param dest A character string (or NULL) for the path within `posts`. By
+#' default, the title is adapted as the directory name.
+#' @param open A logical: have the default editor open a window to edit the
+#' `index.qmd` file?
+#' @param call A call object for reporting errors.
+#' @param ... A named list of values to be added to the yaml header, such as
+#' `description`, `author`, `categories`, etc.
+#' @return The path to the index file.
+#' @export
+#' @examples
+#' \dontrun{
+#'  \donttest{
+#' new_blog_post("making quarto blog posts", categories = c("R"))
+#'
+#'  }
+#' }
+#'
+new_blog_post <- function(title, dest = NULL, open = rlang::is_interactive(),
+                          call = rlang::current_env(), ...) {
+
+  if (is.null(dest)) {
+    # Scrub title to make directory name
+    dest <- gsub("[[:space:]]", "-", tolower(title))
+  }
+  dest_path <- make_post_dir(dest, call)
+  post_yaml <- make_post_yaml(title, ...)
+  qmd_path <- write_post_yaml(post_yaml, dest_path, call)
+  if (open) {
+    utils::file.edit(qmd_path)
+  }
+  invisible(qmd_path)
+}
+
+make_post_dir <- function(dest, call) {
+  working <- fs::path_wd()
+
+  post_path <- fs::path(working, "posts", dest)
+
+  if (fs::dir_exists(post_path)) {
+    cli::cli_abort("There is already a {.code {dest}} directory in 'posts/'",
+                   call = call)
+  } else {
+    ret <- fs::dir_create(post_path)
+  }
+  ret
+}
+
+make_post_yaml <- function(title, ...) {
+  default_values <- list(
+    title = tools::toTitleCase(title),
+    author = tools::toTitleCase(whoami::fullname("Your name")),
+    date = format(Sys.Date(), "%Y-%m-%d"),
+    categories = character(0)
+  )
+
+  user_values <- list(...)
+
+  yml_values <- utils::modifyList(default_values, user_values)
```

Comment:
Probably I forgot about `modifyList` as we have a `quarto:::merge_list()` in the package